core/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -416,6 +416,9 @@ public void updateRoutingEntry(ShardRouting newRouting) throws IOException {
                 // active primaries.
                 throw new IndexShardRelocatedException(shardId(), "Shard is marked as relocated, cannot safely move to state " + newRouting.state());
             }
+            assert newRouting.active() == false || state == IndexShardState.STARTED || state == IndexShardState.RELOCATED ||
```

Comment:
yes, it was relevant before, and actually broken: If we had had this assertion, it would have triggered when we had the test failure here: #23904.